src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,445 @@
+        syntaxKind switch
+        {
+            SyntaxKind.AndAssignmentExpression or
+            SyntaxKind.OrAssignmentExpression => true,
+            _ => false
+        };
+
+    private static SyntaxKind GenerateBinaryExpressionSyntaxKind(SyntaxKind syntaxKind) =>
+        syntaxKind switch
+        {
+            SyntaxKind.AndAssignmentExpression => SyntaxKind.BitwiseAndExpression,
+            SyntaxKind.OrAssignmentExpression => SyntaxKind.BitwiseOrExpression,
+            _ => SyntaxKind.None
+        };
+
+    private static EvaluationResult PreEvaluate(
+        SyntaxNode syntaxNode,
+        ContextInformation contextInformation,
+        Dictionary<SyntaxNode, VariableValue> syntaxNodeToVariableValueMap,
+        bool displayDiagnostics)
+    {
+        var builderToAnalysisContext = contextInformation.BuilderToAnalysisContext;
+        var variableValues = contextInformation.VariableValues;
+        var canEvaluate = false;
+        var syntaxKind = syntaxNode.Kind();
+        var isCompoundExpression = IsCompoundExpression(syntaxKind);
+
+        var rightHandSide = syntaxNode switch
+        {
+            ParenthesizedExpressionSyntax @parenthesizedExpression => @parenthesizedExpression.TrimParenthesis(),
+            AssignmentExpressionSyntax  @assignmentExpression => @assignmentExpression.Right,
+            VariableDeclaratorSyntax @variableDeclarator => @variableDeclarator.Initializer.Value,
+            _ => syntaxNode
+        };
+
+        if (rightHandSide is IdentifierNameSyntax identifierNameSyntax)
+        {
+            canEvaluate = variableValues.TryGetValueSafe(identifierNameSyntax.Identifier.ValueText, out var rightVariableValue);
+
+            if (canEvaluate)
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                syntaxNodeToVariableValueMap.Add(identifierNameSyntax, rightVariableValue);
+
+                if (displayDiagnostics)
+                {
+                    rightVariableValue.Locations.Add(identifierNameSyntax.GetLocation());
+                }
+            }
+        }
+        else if (rightHandSide is InvocationExpressionSyntax invocationExpression)
+        {
+            canEvaluate = builderToAnalysisContext.TryGetValueSafe(invocationExpression, out var expressionAnalysisContext);
+
+            if (canEvaluate)
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                var rightVariableValue = new VariableValue(expressionAnalysisContext, expressionAnalysisContext.Node.Locations.ToList());
+                syntaxNodeToVariableValueMap.Add(invocationExpression, rightVariableValue);
+            }
+        }
+        else if (rightHandSide is BinaryExpressionSyntax binaryExpressionSyntax)
+        {
+            canEvaluate = builderToAnalysisContext.TryGetValueSafe(binaryExpressionSyntax, out var expressionAnalysisContext);
```

Comment:
let's use expressionAnalysisContext as close as possible to TryGetValueSafe.
```
if (builderToAnalysisContext.TryGetValueSafe(binaryExpressionSyntax, out var expressionAnalysisContext))
{ 
 ...
   syntaxNodeToVariableValueMap.Add(binaryExpressionSyntax, rightVariableValue);
   canEvaluate = true;
}
else
{
   leftEvaluationResult
   rightEvaluationResult
}

```

Reply:
Done